crypto/src/traits.rs

```diff
@@ -122,14 +126,44 @@ pub trait Authenticator:
 {
     type PubKey: VerifyingKey<Sig = Self>;
     type PrivKey: SigningKey<Sig = Self>;
+    type AggregateSig: AggregateAuthenticator<Sig = Self>;
 }
 
 /// Trait impl'd by a public / private key pair in asymmetric cryptography.
 ///
-pub trait KeyPair {
+pub trait KeyPair: Sized {
     type PubKey: VerifyingKey<PrivKey = Self::PrivKey>;
     type PrivKey: SigningKey<PubKey = Self::PubKey>;
     fn public(&'_ self) -> &'_ Self::PubKey;
-    fn private(self) -> Self::PrivKey;
+    fn private(&'_ self) -> &'_ Self::PrivKey;
     fn generate<R: CryptoRng + RngCore>(rng: &mut R) -> Self;
+    fn deterministic_generate(seed: &[u8], id: &[u8], domain: &[u8]) -> Result<Self, signature::Error>;
```

Comment:
Some field documentation would help.

The schemes we use do not tie themselves specifically to a KDF solution: do we want to maintain such a tight coupling between a signature scheme and a KDF function?

One possible way to address this would be to KDF generically, there is a trait doing this here for inspiration:
https://github.com/diem/diem/blob/416fdd3b18d02b2a099917467910d5a6865f2fed/crypto/crypto/src/hkdf.rs